tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/mailet/CalDavCollect.java

```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    public static final String SOURCE_ATTRIBUTE_NAME = "source";
+    public static final String DEFAULT_SOURCE_ATTRIBUTE_NAME = "calendars";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CalDavCollect.class);
+    private static final Class<Map<String, AttributeValue<Calendar>>> MAP_STRING_CALENDAR_CLASS = (Class<Map<String, AttributeValue<Calendar>>>) (Object) Map.class;
+
+    private final CalendarClients calendarClients;
+    private DavClient davClient;
+    private DavUserProvider davUserProvider;
+    private AttributeName sourceAttributeName;
+
+    @Inject
+    public CalDavCollect(CalendarClients calendarClients) {
+        this.calendarClients = calendarClients;
+    }
+
+    public void initialize() throws MessagingException {
+        sourceAttributeName = AttributeName.of(getInitParameter(SOURCE_ATTRIBUTE_NAME, DEFAULT_SOURCE_ATTRIBUTE_NAME));
+        davClient = calendarClients.getDavClient().orElseThrow(() ->
+            new MailetException(
+                "Failed to initialize mailet. OpenPaasModule is required to this mailet to function correctly."));
+        davUserProvider = calendarClients.getDavUserProvider().orElseThrow(() ->
+            new MailetException(
+                "Failed to initialize mailet. OpenPaasModule is required to this mailet to function correctly."));
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        initialize();
+
+        try {
+            AttributeUtils.getValueAndCastFromMail(mail, sourceAttributeName, MAP_STRING_CALENDAR_CLASS)
+                .ifPresent(calendars -> calendars.values()
+                    .stream()
+                    .map(AttributeValue::getValue)
+                    .toList()
+                    .forEach(calendar -> handleCalendarInMail(calendar, mail)));
+        } catch (ClassCastException e) {
+            LOGGER.error("Received a mail with {} not being an ICAL object for mail {}", sourceAttributeName, mail.getName(), e);
+        }
+    }
+
+    private void handleCalendarInMail(Calendar calendar, Mail mail) {
+        try {
+            mail.getMaybeSender().asOptional()
+                .ifPresent(sender ->
+                    davUserProvider.provide(Username.of(sender.asString()))
+                        .flatMap(davUser -> davClient.getCalendarObject(davUser, getEventUid(calendar))
+                            .flatMap(davCalendarObject -> updateDavCalendar(davCalendarObject, calendar, davUser))
+                            .switchIfEmpty(createDavCalendar(calendar, davUser)))
+                        .block());
+        } catch (Exception e) {
+            LOGGER.error("Error while handling calendar in mail {}", mail.getName(), e);
+        }
+    }
+
+    private Mono<Void> updateDavCalendar(DavCalendarObject davCalendarObject, Calendar calendar, DavUser davUser) {
+        // TODO Rewrite this method to update the calendar event
```

Comment:
TODO not done spotted?

Reply:
The original ticket only covered CREATE and CANCEL events. 
@hungphan227 is likely want to UPDATE events (e.g., dtstart, dtend, location) in this pr, 
not sure it is necessary or not.

Reply:
If in a following PR, fair!